src/Poller/EventPoller.cpp

```diff
@@ -115,6 +143,13 @@ int EventPoller::addEvent(int fd, int event, PollEventCB cb) {
             _event_map.emplace(fd, std::make_shared<PollEventCB>(std::move(cb)));
         }
         return ret;
+#elif defined(HAS_KQUEUE)
+        struct kevent kev;
+        EV_SET(&kev, fd, toKevent(event), EV_ADD|EV_ENABLE, 0, 0, NULL);
+        kevent(_kqueue_fd, &kev, 1, NULL, 0, NULL);
```

Comment:
Should we refer to epoll here and handle the return value of kevent?

Reply:
Sure. Here is the `kevent`'s man page.

> ```
> RETURN VALUES
>      The kqueue() system call creates a new kernel event queue and returns a file descriptor.  If there was an error creating the kernel event queue, a value of -1 is returned and errno set.
> 
>      The kevent(), kevent64() and kevent_qos() system calls return the number of events placed in the eventlist, up to the value given by nevents.  If an error occurs while processing an
>      element of the changelist and there is enough room in the eventlist, then the event will be placed in the eventlist with EV_ERROR set in flags and the system error in data.  Otherwise,
>      -1 will be returned, and errno will be set to indicate the error condition.  If the time limit expires, then kevent(), kevent64() and kevent_qos() return 0.
> ```

`kevent` api can be used for two purposes: event registering (register event to be notified) & event polling (event happened and be notified). the param `changlist` is used to register event to the kernel, while the `eventlist` is used to let the user space knows what events have emitted, and also the error events during event registering.

For the Delete/Add/Modify operations(Modify == Add), which are the event registering,  because I set the `eventlist` as NULL, so when process the `changelist`, which is the `Delete/Add/Modify` operation, has error, the return val == -1. 

For the event polling, the return value is the number of events in `eventlist`. 

So, return val >= 0 can be consider success, otherwise there is an error.

kqueue只有一个API `kevent`, `EV_SET` 是个宏。事件注册和事件轮询都是同一个 api `kevent()`干的。 `kevent()` 里面的`changlist`就是用来做事件注册的，例如这个例子中的`ADD/DELETE/MODIFY`. 当事件注册时出现错误时，首先会把错误生成一个事件（event)放入`eventlist`里面，但是当`eventlist`没有空间放这个错误时，才会return -1，因为正常情况下`kevent`的返回值是内核通知给用户空间的`eventlist`的数目，也就是数据轮询的数目。

我在这个api使用上，当是`ADD/DELETE/MODIFY`时，把`eventlist`置空，因此此时如果有错误，就会返回-1. 
如此就可以通过判断`kevent()`的返回值>=0 为正常，else 则为错误异常。